### PR TITLE
🔨 Added Annotation Step

### PR DESCRIPTION
**Pull request by migbro**
## Description

Nearly complete updates (waiting on germline snv annot workflow update), but close enough to start a review. This PR:
 - Adds a script to clean up ClinVar vcf, see end of README
 - Adds a bcftools annotation step. README also describes how this works
 - Add the annotation tools repo as a git submodule
 - Update links

Fixes # https://github.com/d3b-center/bixu-tracker/issues/2283

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] [With ClinVar Input](https://cavatica.sbgenomics.com/u/d3b-bixu/d3b-diskin-pathogenicity-dev/tasks/ca5f4452-8b0f-47cd-a4ee-2a1a15d78ac3/)
- [ ] [No ClinVar Input](https://cavatica.sbgenomics.com/u/d3b-bixu/d3b-diskin-pathogenicity-dev/tasks/e1716164-7e4d-402d-a704-6827a25860e3/)

**Test Configuration**:
* Environment:
* Test files:

## Checklist:

- [ x My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings
- [ ] I have committed any related changes to the PR

### CODE REVIEW

**Comment by migbro** [comment]
Just added a test here: https://cavatica.sbgenomics.com/u/d3b-bixu/d3b-diskin-pathogenicity-dev/tasks/7016d1b4-c735-4a42-a6a6-a7b55ef7ee19/ related to a requested feature

File: scripts/cleanup_clinvar.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+            if record.contig in update_dict['old_contigs']:
+                if record.contig != cur:
+                    cur = record.contig
+                    if cur in update_dict['old_contigs']:
+                        rename = update_dict['renamed_contigs'][update_dict['old_contigs'].index(cur)]
+                record.contig = rename
+                # Use variant id in non-canon dict to update alts
+                if record.id in non_canon_dict:
+                    update_alts = update_dict['iupac'][non_canon_dict[record.id]]
+                    split_records(record, update_alts, output)
```
Comment (dmiller15) [inline scripts/cleanup_clinvar.py:47]: so I'm unsure if this is needed. BCFtools doesn't need the records to be decomposed. According to their documentation:
> When multiple ALT alleles are present in the annotation file (given as comma-separated list of alleles), at least one must match one of the alleles in the corresponding VCF record. Similarly, at least one alternate allele from a multi-allelic VCF record must be present in the annotation file.

**Review by dmiller15 — CHANGES_REQUESTED** [review]
First pass thoughts.

File: README.md [hunk 2]
```diff
@@ -75,9 +82,26 @@ The [update_gene_symbols.py](https://github.com/d3b-center/D3b-DGD-Collaboration
...
 Documentation for this can be found [here](docs/INTERVAR_WF.md)
 ### AutoPVS1
 An additional pathogenicity scoring tool, run on the VEP-annotated input.
-Documentation for this can be found [here](https://github.com/d3b-center/D3b-autoPVS1/tree/v2.0.0#readme)
\ No newline at end of file
+Documentation for this can be found [here](https://github.com/d3b-center/D3b-autoPVS1/tree/v2.0.0#readme)
+
+### Optional Inputs
+As mentioned above, the preprocessing workflow can add an additional annotation
+ - `annotation_vcf`: hg38 chromosome-formatted vcf file with multi-allelics split. If provided bcftools will add annotation from the specified columns for each variant that matches
```
Comment (wongjessica93) [inline README.md:95]: If provided, BCFtools will add annotation from the specified columns for each variant that matches

File: README.md [hunk 3]
```diff
@@ -1,14 +1,21 @@
...
+It is recommended to have first run the [Kids First Germline Annotation Workflow](https://github.com/kids-first/kf-annotation-tools/blob/v1.1.0/docs/GERMLINE_SNV_ANNOT_README.md) first.
 
 ## Pathogenicity Preprocessing Workflow
 This workflow uses the prerequisite input to run the InterVar workflow and autoPVS1 tool.
 The major pieces of software being used are:
  - ANNOVAR latest: The software has no versioning, but references do. See `annovar_db` section in [Recommended inputs](#recommended-inputs)
  - InterVar v2.2.1
  - AutoPVS1 v2.0.0: Modified from AutoPVS1 v2.0 to fit annotated KF vcf output. See [README for autoPVS1](https://github.com/d3b-center/D3b-autoPVS1/tree/v2.0.0#readme) for details
+
+Optionally, if you which to add (and in needed, overwrite) another annotation from a vcf file (likely ClinVar), a bcftools strip and annotate steps are provided. The input vcf will be processed and its result will appear as an additional output in the workflow.
```
Comment (wongjessica93) [inline README.md:18]: Optionally, if you wish to add (and overwrite if needed) another annotation from a vcf file (likely ClinVar), a BCFtools strip and annotate steps are provided. The input vcf will be processed and its result will appear as an additional output in the workflow.

File: README.md (hunk 2 above)
Comment (wongjessica93) [inline README.md:95]: - `annotation_vcf`: hg38 chromosome-formatted vcf file with multi-allelics split. If provided, BCFtools will add annotation from the specified columns for each variant that matches

File: README.md [hunk 4]
```diff
@@ -75,9 +82,26 @@ The [update_gene_symbols.py](https://github.com/d3b-center/D3b-DGD-Collaboration
...
+
+### Optional Inputs
+As mentioned above, the preprocessing workflow can add an additional annotation
+ - `annotation_vcf`: hg38 chromosome-formatted vcf file with multi-allelics split. If provided bcftools will add annotation from the specified columns for each variant that matches
+ - `bcftools_annot_columns`: A csv string of from annotation to port into the input vcf. Must provide if `annotation_vcf` given. See [bcftools annotate](https://samtools.github.io/bcftools/bcftools.html#annotate) documentation on how to properly reference
+ - `bcftools_strip_for_vep`: If re-annotating certain `INFO` fields, it's best to strip the old annotation first to avoid conflicts. Use the same format as `bcftools_annot_columns` to reference fields being stripped
+ #### A note on ClinVar annotation
+ For the publication, [ClinVar release 20231028](https://ftp.ncbi.nlm.nih.gov/pub/clinvar/vcf_GRCh38/archive_2.0/2023/clinvar_20231028.vcf.gz) was used. In order to be compatible with our hg38-aligned vcfs, we additionally downloaded the [variant suammry](https://ftp.ncbi.nlm.nih.gov/pub/clinvar/tab_delimited/variant_summary.txt.gz) file, ran a [custom script](scripts/cleanup_clinvar.py) that:
```
Comment (wongjessica93) [inline README.md:100]: For the publication, [ClinVar release 20231028](https://ftp.ncbi.nlm.nih.gov/pub/clinvar/vcf_GRCh38/archive_2.0/2023/clinvar_20231028.vcf.gz) was used. In order to be compatible with our hg38-aligned vcfs, we additionally downloaded the [variant summary]

File: README.md [hunk 5]
```diff
@@ -1,14 +1,21 @@
...
-
+To run, please see the [CAVATICA app](https://cavatica.sbgenomics.com/public/apps/cavatica/apps-publisher/d3b-diskin-pathogenicity-preprocess-wf). Each version should correspond with a git release. This repo makes use of the git submodule feature for ease of code maintenance. To properly retrieve all relevant code:
+```sh
+git clone  https://github.com/d3b-center/D3b-Pathogenicity-Preprocessing
+git submodule init
+git submodule update
+```
 ## Prequisite
-It is recommended to have first run the [Kids First Germline Annotation Workflow](https://github.com/kids-first/kf-germline-workflow/blob/v1.0.0/docs/GERMLINE_SNV_ANNOT_README.md) first.
+It is recommended to have first run the [Kids First Germline Annotation Workflow](https://github.com/kids-first/kf-annotation-tools/blob/v1.1.0/docs/GERMLINE_SNV_ANNOT_README.md) first.
```
[the remark this answers is not in the document]
Comment (migbro) [reply README.md:9]: Um, tough. I am thinking/hoping it's unlikely that for this workflow, future releases will impact it....and if it does, we should point to the new version and update this one

File: README.md [hunk 6]
```diff
@@ -75,9 +82,26 @@ The [update_gene_symbols.py](https://github.com/d3b-center/D3b-DGD-Collaboration
...
 With `results` used to replace `PVS1.level` file. Recommend references for this workflow can be obtained [here](https://cavatica.sbgenomics.com/u/kfdrc-harmonization/kf-references/files/#q?path=d3b_diskin_pathogenicity).
+
 ### InterVar Classification Workflow
 This workflow is a critical component in generating scoring metrics needed to classify pathogenicity of variants.
 Documentation for this can be found [here](docs/INTERVAR_WF.md)
 ### AutoPVS1
 An additional pathogenicity scoring tool, run on the VEP-annotated input.
-Documentation for this can be found [here](https://github.com/d3b-center/D3b-autoPVS1/tree/v2.0.0#readme)
\ No newline at end of file
+Documentation for this can be found [here](https://github.com/d3b-center/D3b-autoPVS1/tree/v2.0.0#readme)
```
[the remark this answers is not in the document]
Comment (migbro) [reply README.md:91]: Nah. Master occasionally has a way of not being what you want

**Review by dmiller15 — APPROVED** [review]

**Review by wongjessica93 — APPROVED** [review]

**Pull request merged**